src/dhcp/aca_dhcp_server.cpp

```diff
@@ -0,0 +1,735 @@
+    ACA_LOG_INFO("Entry not exist!  (mac = %s)\n", dhcp_cfg_in->mac_address.c_str());
+    return EXIT_SUCCESS;
+  }
+
+  _dhcp_db_mutex.lock();
+  _dhcp_db->erase(dhcp_cfg_in->mac_address);
+  _dhcp_db_mutex.unlock();
+
+  return EXIT_SUCCESS;
+}
+
+int ACA_Dhcp_Server::update_dhcp_entry(dhcp_config *dhcp_cfg_in)
+{
+  //dhcp_entry_data stData = {0};
+  std::unordered_map<string, dhcp_entry_data>::iterator pos;
+  dhcp_entry_data *pData = nullptr;
+
+  if (_validate_dhcp_entry(dhcp_cfg_in)) {
+    ACA_LOG_ERROR("Valiate dhcp cfg failed! (mac = %s)\n",
+                  dhcp_cfg_in->mac_address.c_str());
+    return EXIT_FAILURE;
+  }
+
+  pData = _search_dhcp_entry(dhcp_cfg_in->mac_address);
+  if (nullptr == pData) {
+    ACA_LOG_ERROR("Entry not exist! (mac = %s)\n", dhcp_cfg_in->mac_address.c_str());
+    return EXIT_FAILURE;
+  }
+
+  _dhcp_db_mutex.lock();
+  DHCP_ENTRY_DATA_SET(pData, dhcp_cfg_in);
+  _dhcp_db_mutex.unlock();
+
+  return EXIT_SUCCESS;
+}
+
+dhcp_entry_data *ACA_Dhcp_Server::_search_dhcp_entry(string mac_address)
+{
+  std::unordered_map<string, dhcp_entry_data>::iterator pos;
+
+  pos = _dhcp_db->find(mac_address);
+  if (_dhcp_db->end() == pos) {
+    return nullptr;
+  }
+
+  return (dhcp_entry_data *)&(pos->second);
+}
+
+void ACA_Dhcp_Server::_validate_mac_address(const char *mac_string)
+{
+  unsigned char mac[6];
+
+  if (mac_string == nullptr) {
+    throw std::invalid_argument("Input mac_string is null");
+  }
+
+  if (sscanf(mac_string, "%hhx:%hhx:%hhx:%hhx:%hhx:%hhx", &mac[0], &mac[1],
+             &mac[2], &mac[3], &mac[4], &mac[5]) == 6) {
+    return;
+  }
+
+  if (sscanf(mac_string, "%hhx-%hhx-%hhx-%hhx-%hhx-%hhx", &mac[0], &mac[1],
+             &mac[2], &mac[3], &mac[4], &mac[5]) == 6) {
+    return;
+  }
+
+  // nothing matched
+  ACA_LOG_ERROR("Invalid mac address: %s\n", mac_string);
+
+  throw std::invalid_argument("Input mac_string is not in the expect format");
+}
+
+void ACA_Dhcp_Server::_validate_ipv4_address(const char *ip_address)
+{
+  struct sockaddr_in sa;
+
+  // inet_pton returns 1 for success 0 for failure
+  if (inet_pton(AF_INET, ip_address, &(sa.sin_addr)) != 1) {
+    throw std::invalid_argument("Virtual ipv4 address is not in the expect format");
+  }
+}
+
+void ACA_Dhcp_Server::_validate_ipv6_address(const char *ip_address)
+{
+  struct sockaddr_in sa;
+
+  // inet_pton returns 1 for success 0 for failure
+  if (inet_pton(AF_INET6, ip_address, &(sa.sin_addr)) != 1) {
+    throw std::invalid_argument("Virtual ipv6 address is not in the expect format");
+  }
+}
+
+int ACA_Dhcp_Server::_validate_dhcp_entry(dhcp_config *dhcp_cfg_in)
+{
+  if (0 >= dhcp_cfg_in->mac_address.size()) {
+    throw std::invalid_argument("Input mac_string is null");
+  }
+
+  _validate_mac_address(dhcp_cfg_in->mac_address.c_str());
+
+  if (0 < dhcp_cfg_in->ipv4_address.size()) {
+    _validate_ipv4_address(dhcp_cfg_in->ipv4_address.c_str());
+  }
+
+  if (0 < dhcp_cfg_in->ipv6_address.size()) {
```

Comment:
@xieus do we have ipv6 address assigned to a new port currently, and will we always assign one in the final implementation?

@w2520n2520 once we have confirmation, you may need to remove the ipv6 requirement or check here.

Reply:
If no ipv6 address assigned, so the field in goal message will be not initialized or absent?

Reply:
if no ipv6 assigned, ipv6_address will be empty. Looks like your check for size > 0 will handle it just find. Sorry I miss that.